src/Package/Impl/DataInspect/Viewers/GridViewerBase.cs

```diff
@@ -6,35 +6,35 @@
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Common.Core.Services;
-using Microsoft.Common.Core.Shell;
 using Microsoft.Common.Core.Threading;
+using Microsoft.R.Components.Settings;
 using Microsoft.R.DataInspection;
-using Microsoft.VisualStudio.R.Package.Shell;
 using Microsoft.VisualStudio.R.Package.Utilities;
 using Microsoft.VisualStudio.Shell.Interop;
 using static Microsoft.R.DataInspection.REvaluationResultProperties;
+using static System.FormattableString;
 
 namespace Microsoft.VisualStudio.R.Package.DataInspect.Viewers {
     internal abstract class GridViewerBase : ViewerBase, IObjectDetailsViewer {
         private const int _toolWindowIdBase = 100;
         private const REvaluationResultProperties _properties =
            ClassesProperty | ExpressionProperty | TypeNameProperty | DimProperty | LengthProperty;
 
-        private readonly IServiceContainer _services;
-
-        public GridViewerBase(IServiceContainer services, IDataObjectEvaluator evaluator) : base(evaluator) {
-            _services = services;
-        }
+        protected GridViewerBase(IServiceContainer services, IDataObjectEvaluator evaluator) : 
+            base(services, evaluator) { }
 
         #region IObjectDetailsViewer
         public ViewerCapabilities Capabilities => ViewerCapabilities.List | ViewerCapabilities.Table;
 
         public abstract bool CanView(IRValueInfo evaluation);
 
         public async Task ViewAsync(string expression, string title, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (!Services.GetService<IRSettings>().GridDynamicEvaluation) {
+                expression = Invariant($"as.data.frame({expression})");
```

Comment:
Why convert to data frame?

Reply:
Works for #3772

Reply:
But then why do it conditional on `GridDynamicEvaluation`?

More importantly, converting to a data frame is a lossy operation for matrices and a bunch of other stuff. The code in `grid.R` actually goes to great lengths to avoid doing this for that exact reason (originally, it did such a conversion, and I kept finding corner cases where it would do the wrong thing). 

We need to figure out what exactly makes tibble different for sorting purposes, and adjust `grid.R` accordingly, to fix this right. The docs for tibble seem to indicate that `arrange()` is the right function for the job, but it would require substantial rework of our code to make use of it. 

If a quick fix is needed right now, then let's do `as.data.frame` in grid.R only for tibbles.

Reply:
(It looks like the easiest way to do the type check without having to load the tibble library is to check for `"tbl"` in `class()`)

Reply:
I wanted to avoid changing dynamic case since it is tailored to large sets specifically. We can try check for tibbles, it gets more complicated though since row selector expression is passed separately and just converting `x` in grid.R doesn't work.

Reply:
One thing i want to check before investing more is to get telemetry on how many uses do we get on dynamic mode. RS seems to be happy without it?

Reply:
RStudio data grid is dynamic by default, actually. It just doesn't capture all changes as reliably as we do - they seem to be trying to refresh it only when they see `<-` in REPL input, based on my observarions, while we refresh on eveery command. So if you define a function that changes your data, and then later invoke that function, the grid doesn't refresh. But e.g. try this:
```r
mtcars <- mtcars
View(mtcars)
mtcars[[1, 1]] <- mtcars[[1, 1]] + 1
```

Reply:
Converting `x` first thing in `grid_data` ought to work. The row selector is passed as a function, and that function accepts the dataset as the argument, which is passed by `grid_data` (I believe we use `x` as the name for that argument when we generate the call to `grid_data`, but it's not the same `x` as `grid_data` itself accepts). So if you change `x`, the row selector should just get the changed value as the argument, and apply its filtering logic to that - which should work as expected with a data frame.

Reply:
OK. This prob will render unnecessary first part specific to 1D